File: C/pancakeSort.c
// Pancake sort in C
#include <stdlib.h> 
#include <stdio.h> 
void flip(int arr[] , int i){


	for(int j = 0 ; j <= (i / 2) ; j++)
	{
		int temp = arr[i-j] ; 
		arr[i-j] = arr[j] ; 
		arr[j] = temp ; 
	}
}
// time complexity = O(n) ; Space complexity = O(1) 

int max_element(int arr[] , int n){
    int maks = arr[0] ;
    int temp = 0 ; 
    for(int i = 1 ; i < n ; i++)
    {
        if(maks < arr[i])
        {
            maks = arr[i] ; 
            temp = i ; 
        }
    }
    return temp ; 
}

void sort(int arr[] , int n){
	int curr_size = n ; 
	while(curr_size > 0)
	{
		int max_index = max_element(arr , curr_size) ; // time complexity = O(n) ; space complexity = O(1);
		if(max_index !=curr_size -1)
		{
		flip(arr, max_index); 
		flip(arr, curr_size -1) ;}
		curr_size-- ; 
		
	}
}

//time complexity = O(n * max(p, q)) where p is the time complexity of flip() and q is the complexity of max_element(); space complexity = O(1)

int main()
{


	int n ; (void)scanf("%d", &n);   
	int a[n] ; 
	for(int i = 0 ; i < n ; i ++)
	{
		(void)scanf("%d", &a[i]);  
	}


	sort(a , n); 
	for(int i = 0 ; i < n ; i ++)
	{
    	printf("%d ",a[i]);
	}


}
/* here n is the number of elements in the array to be sorted and a[0...n-1] is the array.

input example:
5
5 4 3 2 1 
Output:
1 2 3 4 5

Overall time complexity = O(n) ; Overall Space complexity = O(1) */

File: Python/PancakeSort.py
def flip(array, j):
    start = 0
    while(start<j):
        temp = array[start]
        array[start] = array[j]
        array[j] = temp
        start = start +1
        j -= 1

def findMaxIdx(array, n):
    mi = 0
    for i in range(n):
        if (array[i] > array[mi]):
            mi = i
    return mi
    
n = int(input("Enter number of elements:"))
print("Enter elements:")
array = []
for i in range(n):
    array.append(int(input()))
x = n
while (x>1):
    j = findMaxIdx(array, x)
    if (j != x-1):
        flip(array, j)
        flip(array, x-1)
    x -= 1
    
print("Sorted array is:")
print(array)

"""
Pancake Sort in python.

Sample I/O:

Enter number of elements:5
Enter elements:
12
43
13
56
2
Sorted array is:
[2, 12, 13, 43, 56]

Enter number of elements:4
Enter elements:
45
44
43
1
Sorted array is:
[1, 43, 44, 45]

Time Complexity: O(n^2)
Space Complexity: O(1)

"""
